ocaml/xapi/extauth_plugin_ADwinbind.ml

```diff
@@ -34,6 +38,118 @@ module Winbind = struct
+
+let config_winbind_damon ~domain ~workgroup =
+  let open Xapi_stdext_unix in
+  let smb_config = "/etc/samba/smb.conf" in
+  let conf_contents =
+    String.concat "\n"
+      [
+        "# autogenerated by xapi"
+      ; "[global]"
+      ; "kerberos method = secrets and keytab"
+      ; Printf.sprintf "realm = %s" domain
+      ; "security = ADS"
+      ; "template shell = /bin/bash"
+      ; "winbind offline logon = Yes"
+      ; "winbind refresh tickets = Yes"
+      ; "winbind enum groups = no"
+      ; "winbind enum users = no"
+      ; Printf.sprintf "workgroup = %s" workgroup
+      ; "idmap config * : range = 3000000-3999999"
+      ; Printf.sprintf "idmap config %s: backend = rid" domain
+      ; Printf.sprintf "idmap config %s: range = 2000000-2999999" domain
+      ; "idmap config * : backend = tdb"
+      ; "" (* Empty line at the end *)
+      ]
+  in
+  let len = String.length conf_contents in
+  Unixext.atomic_write_to_file smb_config 0o0644 (fun fd ->
+      let (_ : int) = Unix.single_write_substring fd conf_contents 0 len in
+      ())
+
+let from_config ~name ~err_msg ~config_params =
+  match List.assoc_opt name config_params with
+  | Some v ->
+      v
+  | _ ->
+      raise (Auth_service_error (E_GENERIC, err_msg))
+
+let assert_hostname_valid () =
+  let hostname =
+    (fun __context ->
+      Helpers.get_localhost ~__context |> fun host ->
+      Db.Host.get_hostname ~__context ~self:host)
+    |> Server_helpers.exec_with_new_task "retrieving hostname"
+  in
+  let all_numbers hostname =
+    Re.Perl.re {|^\d+$|} |> Re.Perl.compile |> fun re ->
+    Re.matches re hostname <> []
+  in
+  if all_numbers hostname then
+    raise
+      (Auth_service_error
+         ( E_GENERIC
+         , Printf.sprintf "hostname '%s' cannot contain only digits." hostname
+         ))
+
+let assert_domain_equal_service_name ~service_name ~config_params =
+  (* For legeacy support, if domain exist in config_params, it must be equal to service_name *)
+  let domain_key = "domain" in
+  match List.assoc_opt domain_key config_params with
+  | Some domain ->
+      if domain <> service_name then
+        raise
+          (Auth_service_error
+             (E_GENERIC, "if present, config:domain must match service-name."))
+  | None ->
+      ()
+
+let extract_ou_config ~config_params =
+  try
+    let ou = from_config ~name:"ou" ~err_msg:"" ~config_params in
+    ([("ou", ou)], [Printf.sprintf "createcomputer=%s" ou])
+  with Auth_service_error _ -> ([], [])
+
+let clean_machine_account ~service_name = function
+  | Some u, Some p -> (
+      (* Clean machine account in DC *)
+      let env = [|Printf.sprintf "PASSWD=%s" p|] in
+      let args = ["ads"; "leave"; "-U"; u] in
+      try
+        Helpers.call_script ~env net_cmd args |> ignore ;
+        debug "Clean machine account in domain %s successfully" service_name
+      with _ ->
+        let msg =
+          Printf.sprintf "Failed to clean machine account in domain  %s"
```

Comment:
Maybe "Failed the clean machine count" as "clean machine count" is not a verb" or "failed to perform a …"

Reply:
`machine account` is none,  does `clean the machine account` make sense to you?

Reply:
Sorry, I had misread that but I'm still not sure what the command does. Is it cleaning the machine account? If so, it "Cleaning the machine account succeeded", "Cleaning the machine account failed" would make sense to me.

Reply:
The command will leave domain, and it will clean the machine account in Domain Controller for us.  (and destroy the machine password)

Reply:
updated